### PR TITLE
refactor(pep440): ensure strict null check

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes:

- Fix strict null errors.

## Context:

- Ref #12869 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: .ls-lint.yml [hunk 1]
```diff
@@ -17,3 +17,4 @@ ignore:
   - node_modules
   - SECURITY.md
   - test/e2e/node_modules
+  - lib/types/renovate_pep440.d.ts
```
Comment (rarkins) [inline .ls-lint.yml:20]: Does it need to be this file name?

Comment (RahulGautamSingh) [reply .ls-lint.yml:20]: No, the name isn't important.

Comment (viceice) [reply .ls-lint.yml:20]: We should move the types to https://github.com/renovatebot/pep440 now, before we merge this PR.

Comment (RahulGautamSingh) [reply .ls-lint.yml:20]: I have created a PR for the same [link](https://github.com/renovatebot/pep440/pull/113)

**Comment by RahulGautamSingh** [comment]
Blocked by [renovate/pep440(#113)](https://github.com/renovatebot/pep440/pull/113)

**Review by viceice — COMMENTED** [review]
blocked by https://github.com/renovatebot/pep440/pull/122

**Comment by viceice** [comment]
You can continue on this pr, you need to use this package now https://www.npmjs.com/package/@renovatebot/pep440

**Comment by RahulGautamSingh** [comment]
@viceice Ready for review and merge.🟩

File: lib/versioning/pep440/range.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
-import { gte, lt, lte, satisfies } from '@renovate/pep440';
-import { parse as parseRange } from '@renovate/pep440/lib/specifier.js';
-import { parse as parseVersion } from '@renovate/pep440/lib/version.js';
+import { gte, lt, lte, satisfies } from '@renovatebot/pep440';
+import { parse as parseRange } from '@renovatebot/pep440/lib/specifier';
+import { parse as parseVersion } from '@renovatebot/pep440/lib/version';
```
Comment (viceice) [inline lib/versioning/pep440/range.ts:3]:
```suggestion
import { parse as parseRange } from '@renovatebot/pep440/lib/specifier.js';
import { parse as parseVersion } from '@renovatebot/pep440/lib/version.js';
```

😕  We should also export those from index. `.js` is required for `ESM` compabillity

Comment (viceice) [reply lib/versioning/pep440/range.ts:3]: new exports can be done in a later PR

File: lib/versioning/pep440/index.ts [hunk 3]
```diff
@@ -31,27 +31,38 @@ const isStable = (input: string): boolean => {
...
   return found.length === 0 ? null : found[0];
-};
+}
 
-export const isSingleVersion = (constraint: string): string =>
-  isVersion(constraint) ||
-  (constraint?.startsWith('==') && isVersion(constraint.substring(2).trim()));
+export function isSingleVersion(constraint: string): boolean {
+  return (
+    !!isVersion(constraint) ||
```
[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply lib/versioning/pep440/index.ts:56]: https://github.com/renovatebot/renovate/blob/574196a384a4460bd6eaec57660a195a29badd23/lib/versioning/pep440/index.ts#L12-L23
We are importing `valid` from pep440 as `isVersion`.
https://github.com/renovatebot/pep440/blob/4d2dddbe07989cbb75963cf1d038f9829fc573d0/lib/version.d.ts#L8
And inside the pep440 implementation `valid` returns a `string|null`, so to change the return type of `isVersion` we have to define a new function.

[the remark this answers is not in the document]
Comment (viceice) [reply lib/versioning/pep440/index.ts:56]: ok, leave it for now

File: lib/versioning/pep440/index.ts [hunk 4]
```diff
@@ -22,6 +21,9 @@ const {
   eq,
 } = pep440;
 
+const isVersion = (input: string): boolean => {
+  return !!valid(input);
+};
```
Comment (viceice) [inline lib/versioning/pep440/index.ts:26]:
```suggestion
function isVersion(input: string): boolean {
  return !!valid(input);
};
```

File: lib/versioning/pep440/index.ts [hunk 5]
```diff
@@ -31,27 +33,37 @@ const isStable = (input: string): boolean => {
...
+    isVersion(constraint) ||
+    (constraint?.startsWith('==') && isVersion(constraint.substring(2).trim()))
+  );
+}
 
 export { isVersion, matches };
 
 const equals = (version1: string, version2: string): boolean =>
-  isVersion(version1) && isVersion(version2) && eq(version1, version2);
+  !!isVersion(version1) && !!isVersion(version2) && eq(version1, version2);
```
Comment (viceice) [inline lib/versioning/pep440/index.ts:66]:
```suggestion
  isVersion(version1) && isVersion(version2) && eq(version1, version2);
```

**Review by viceice — COMMENTED** [review]
just some small issues left

File: lib/versioning/pep440/index.ts [hunk 6]
```diff
@@ -22,6 +21,9 @@ const {
   eq,
 } = pep440;
 
+function isVersion(input: string): boolean {
+  return !!valid(input);
+};
```
Comment (viceice) [inline lib/versioning/pep440/index.ts:26]:
```suggestion
}
```

File: lib/versioning/pep440/index.ts [hunk 7]
```diff
@@ -22,6 +21,9 @@ const {
   eq,
 } = pep440;
 
+function isVersion(input: string): boolean {
+  return !!valid(input);
+}
```
Comment (viceice) [inline lib/versioning/pep440/index.ts:26]:
```suggestion
}

```

**Review by viceice — APPROVED** [review]

**Pull request merged**